subsys/debug/tracing/tracing_format_sync.c

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2019 Intel corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <tracing_core.h>
+#include <tracing_buffer.h>
+#include <tracing_format_common.h>
+
+void tracing_format_string(const char *str, ...)
+{
+	u8_t *data;
+	va_list args;
+	bool put_success;
+	u32_t length, tracing_buffer_size;
+
+	if (!is_tracing_enabled()) {
+		return;
+	}
+
+	tracing_buffer_size = tracing_buffer_capacity_get();
+
+	va_start(args, str);
+
+	TRACING_LOCK();
+	put_success = tracing_format_string_put(str, args);
+
+	if (put_success) {
+		length = tracing_buffer_get_claim(&data, tracing_buffer_size);
+		tracing_buffer_handle((u8_t *)data, length);
+		tracing_buffer_get_finish(length);
+	} else {
+		tracing_packet_drop_handle();
+	}
+	TRACING_UNLOCK();
+
+	va_end(args);
+}
+
+void tracing_format_raw_data(const char *data, u32_t length)
+{
+	if (!is_tracing_enabled()) {
+		return;
+	}
+
+	TRACING_LOCK();
+	tracing_buffer_handle((u8_t *)data, length);
+	TRACING_UNLOCK();
+}
+
+void tracing_format_data(tracing_data_t *tracing_data_array, u32_t count)
+{
+	u8_t *data;
+	bool put_success;
+	u32_t length, tracing_buffer_size;
+
+	if (!is_tracing_enabled()) {
+		return;
+	}
+
+	tracing_buffer_size = tracing_buffer_capacity_get();
+
+	TRACING_LOCK();
+	put_success = tracing_format_data_put(tracing_data_array, count);
+
+	if (put_success) {
+		length = tracing_buffer_get_claim(&data, tracing_buffer_size);
+		tracing_buffer_handle((u8_t *)data, length);
+		tracing_buffer_get_finish(length);
+	} else {
+		tracing_packet_drop_handle();
```

Comment:
> Looks like 2 of the 3 are copying data into a buffer and immediately pulling it back out for handling simply because it needs to be formatted into "data" and "length" parameters.

yes, because tracing_buffer_handle need them.

 > Is there a different non-buffered method like tracing_format_raw_data()?

I just define 3 different APIs from which end users have to choose based on their usage cases. In this file they work in sync way.